test/common/upstream/cds_speed_test.cc

```diff
@@ -0,0 +1,191 @@
+      envoy::config::endpoint::v3::ClusterLoadAssignment cluster_load_assignment;
+      cluster_load_assignment.set_cluster_name("fare_" + std::to_string(i));
+
+      auto* endpoints = cluster_load_assignment.add_endpoints();
+      endpoints->set_priority(1);
+      auto* locality = endpoints->mutable_locality();
+      locality->set_region("region");
+      locality->set_zone("zone");
+      locality->set_sub_zone("sub_zone");
+
+      auto* resource = response->mutable_resources()->Add();
+      resource->PackFrom(cluster_load_assignment);
+      if (v2_config_) {
+        RELEASE_ASSERT(resource->type_url() ==
+                           "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment",
+                       "");
+        resource->set_type_url("type.googleapis.com/envoy.api.v2.ClusterLoadAssignment");
+      }
+    }
+    validation_visitor_.setSkipValidation(ignore_unknown_dynamic_fields);
+
+    state_.SetComplexityN(num_clusters);
+    state_.ResumeTiming();
+    grpc_mux_->grpcStreamForTest().onReceiveMessage(std::move(response));
+  }
+
+  TestDeprecatedV2Api _deprecated_v2_api_;
+  State& state_;
+  const bool v2_config_;
+  const std::string type_url_;
+  uint64_t version_{};
+  bool initialized_{};
+  Stats::IsolatedStoreImpl stats_;
+  Config::SubscriptionStats subscription_stats_;
+  Ssl::MockContextManager ssl_context_manager_;
+  envoy::config::cluster::v3::Cluster eds_cluster_;
+  NiceMock<MockClusterManager> cm_;
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  EdsClusterImplSharedPtr cluster_;
+  Config::SubscriptionCallbacks* eds_callbacks_{};
+  Config::OpaqueResourceDecoderImpl<envoy::config::endpoint::v3::ClusterLoadAssignment>
+      resource_decoder_{validation_visitor_, "cluster_name"};
+  NiceMock<Random::MockRandomGenerator> random_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Server::MockAdmin> admin_;
+  Singleton::ManagerImpl singleton_manager_{Thread::threadFactoryForTest()};
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  ProtobufMessage::MockValidationVisitor validation_visitor_;
+  Api::ApiPtr api_;
+  Grpc::MockAsyncClient* async_client_;
+  NiceMock<Grpc::MockAsyncStream> async_stream_;
+  Config::GrpcMuxImplSharedPtr grpc_mux_;
+  Config::GrpcSubscriptionImplPtr subscription_;
+};
+
+} // namespace Upstream
+} // namespace Envoy
+
+static void addClusters(State& state) {
+  Envoy::Thread::MutexBasicLockable lock;
+  Envoy::Logger::Context logging_state(spdlog::level::warn,
+                                       Envoy::Logger::Logger::DEFAULT_LOG_FORMAT, lock, false);
+  for (auto _ : state) { // NOLINT(clang-analyzer-deadcode.DeadStores)
+    Envoy::Upstream::CdsSpeedTest speed_test(state, state.range(0));
+    // if we've been instructed to skip tests, only run once no matter the argument:
+    uint32_t clusters = skipExpensiveBenchmarks() ? 1 : state.range(2);
+    speed_test.clusterHelper(state.range(1), clusters);
```

Comment:
Wouldn't this also include the time it took to build the response proto in the benchmark?

Reply:
I don't believe so; the calls to ```state_.PauseTiming()``` and ```ResumeTiming()``` in clusterHelper should exclude everything but the grpc callback?

Reply:
It would be clearer I think what we were benchmarking if we put the PauseTiming calls at the start of the loop here (line 162) and Resume at the end, and then put Resume/Pause calls in the helper functions surrounding the code we really want to test.

Reply:
I think this is what you mean.